File: rust_tradingengine/src/domain/model/ack.rs
use serde::{Serialize, Deserialize};
use crate::domain::enums::{OrderSide, OrderStatus, OrderType, TimeInForce};

/// Represents an order acknowledgment from the exchange
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Ack {
    /// Exchange order ID
    pub order_id: String,
    
    /// Client-assigned order ID
    pub client_order_id: Option<u64>,
    
    /// Name of the instrument
    pub instrument_name: String,
    
    /// Order direction (buy/sell)
    pub direction: OrderSide,
    
    /// Order price (if limit order)
    pub price: Option<f64>,
    
    /// Total order amount
    pub amount: f64,
    
    /// Amount that has been filled
    pub filled_amount: f64,
    
    /// Amount still in the order book
    pub remaining_amount: f64,
    
    /// Current order status
    pub status: OrderStatus,
    
    /// Order type (limit, market)
    pub order_type: OrderType,
    
    /// Time in force setting
    pub time_in_force: TimeInForce,
    
    /// Reason for order status change
    pub change_reason: String,
    
    /// Reason for order deletion (if applicable)
    pub delete_reason: Option<String>,
    
    /// Reason for order insertion (if applicable)
    pub insert_reason: Option<String>,
    
    /// Order creation timestamp (seconds since epoch)
    pub create_time: f64,
    
    /// Whether order is persistent
    pub persistent: bool,
    
    /// Timestamp when record was processed by Rust system (seconds since epoch)
    pub processing_timestamp: Option<f64>,
}

impl Ack {
    /// Add processing timestamp to the Ack
    pub fn with_processing_timestamp(mut self) -> Self {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();
            
        self.processing_timestamp = Some(now);
        self
    }
}

File: python_pipeline/src/python_pipeline/models/ack.py
"""
Ack data model for CrypticsLabBot.
Implements Pydantic models for order acknowledgment data.
"""

from __future__ import annotations

import datetime
from enum import Enum
from typing import ClassVar, Literal, Optional

from pydantic import Field

from python_pipeline.models.model_base import ModelBase


class OrderSide(str, Enum):
    """Order side enum (buy/sell)"""
    BUY = "buy"
    SELL = "sell"


class OrderStatus(str, Enum):
    """Order status enum"""
    OPEN = "open"
    PARTIALLY_FILLED = "partially_filled"
    CANCELLED = "cancelled"
    CANCELLED_PARTIALLY_FILLED = "cancelled_partially_filled"
    FILLED = "filled"


class OrderType(str, Enum):
    """Order type enum"""
    LIMIT = "limit"
    MARKET = "market"


class TimeInForce(str, Enum):
    """Time in force enum"""
    GOOD_TILL_CANCELLED = "good_till_cancelled"
    IMMEDIATE_OR_CANCEL = "immediate_or_cancel"


class ThalexAck(ModelBase):
    """Ack model based on the Thalex order acknowledgment structure"""
    
    # Class variables for the ModelBase
    model_name: ClassVar[str] = "ack"
    
    # Fields matching exact structure needed for the database
    order_id: str = Field(..., description="Exchange order ID")
    client_order_id: Optional[int] = Field(None, description="Client order ID")
    instrument_name: str = Field("BTC-PERPETUAL", description="Instrument name")
    direction: OrderSide = Field(..., description="Order direction (buy/sell)")
    price: Optional[float] = Field(None, description="Order price")
    amount: float = Field(..., description="Order amount")
    filled_amount: float = Field(..., description="Amount filled")
    remaining_amount: float = Field(..., description="Remaining amount")
    status: OrderStatus = Field(..., description="Order status")
    order_type: OrderType = Field(..., description="Order type (limit/market)")
    time_in_force: TimeInForce = Field(..., description="Time in force")
    change_reason: str = Field(..., description="Reason for change")
    delete_reason: Optional[str] = Field(None, description="Reason for deletion")
    insert_reason: Optional[str] = Field(None, description="Reason for insertion")
    create_time: float = Field(..., description="Creation timestamp")
    persistent: bool = Field(..., description="Is order persistent")

